Added support for hammerblade stats tag and printing

marcusnchow opened the pull request:
DRAMsim3 will now print out stats with corresponding tags from hammerblade

marcusnchow committed:
Added support for hammerblade stats tag and printing

dpetrisko commented on src/controller.cc:
```diff
@@ -294,6 +294,18 @@ void Controller::PrintEpochStats() {
     return;
 }
 
+void Controller::PrintEpochStats(int tag) {
```
What's the advantage of using a tag rather than automatically incrementing the epoch?

marcusnchow replied:
The tag is used in hammerblade's vanilla stats profiler, so the point of this commit is to align dramsim printstats, with hammerblade. we dont want to have each epoch have its own tag

dpetrisko commented on src/dram_system.cc:
```diff
@@ -163,9 +182,9 @@ void JedecDRAMSystem::ClockTick() {
-    if (clk_ % config_.epoch_period == 0) {
-        PrintEpochStats();
-    }
+//    if (clk_ % config_.epoch_period == 0) {
```
Do we need to break the epoch_period functionality? Maybe PrintTagStats rather than PrintEpochStats since it's somewhat orthogonal?

drichmond replied:
Let's keep periodic printing, but emit the tag stats to a different file.

@marcusnchow is that doable?

drichmond replied:
Maintain old functionality, append new functionality
(It will mean two .json files, but the alternative is turning off the old file -- which is fine too)

marcusnchow replied:
yeah, we can add an option to turn of the old file

marcusnchow replied:
i turned it off because there are two statsprint epoch period, one from hammerblade and one from dramsim, so with this dramsim prints when hammerblade prints

drichmond replied:
What about the two file approach?

marcusnchow replied:
ye that should work as well

drichmond replied:
I think dan's point about not breaking old functionality is a good one. I don't mind if the old method has a way to be turned off, but I think having a second file for our a stats is a good idea

marcusnchow replied:
wait,,,, is it that you want the specific epoch period from dramsim3? or can using the epoch period from manycore work?  right now, it still prints out periodically, only it gets the period signal from manycore. Otherwise we would need to keep track of two separate printing periods. Do we want to duplicate it?

drichmond replied:
Maybe I'm not following. The manycore stats are printed when the profiler calls are made, and your work added that to dramsim3.

marcusnchow replied:
omg, no im just loosing it, i understand what your saying now, i know how i can do that know

marcusnchow committed:
Split tag and epoch stats

The goal of this commit is to align Dramsim3 stats logging with vanilla_stats in
the manycore

This commit creates two sets of Counters (Epoch and Tag) and associated Printing
Functions (PrintEpochStats and PrintTagStats) within each class
memory_system->dram_system->controller->simple_stats

Within each set there are three types:
{set}_counters_ {set}_vec_counters {set}_histo_counts_

The only difference between the two sets of Counters is the period when they
print to a file and reset their counters

Epoch - prints every "epoch_period" cycle which is set in each dramsim3 config
file after it prints a period, the stats are accumulated in the total counter
and then reset to 0

Tag - prints whenever "print_stat_v_i" in manycore_tb_top is set high, along
with the associated tag in "print_stat_tag_i". This is the same with how
vanilla_stats.csv gets logged. After it prints the counter is reset to 0, but
they are not accumulated.

marcusnchow commented:
alright, i split the epoch stats and tag stats, so now there are two .json files that get printed out

drichmond commented:
This looks good to me. @mrutt92 what say you?

drichmond commented:
@marcusnchow Can you add a few comments (or a meta-comment) about what you did? It seems like you have a naming structure here so just describe what your added methods do for posterity

marcusnchow commented:
Added this to the commit message
```
The goal of this commit is to align Dramsim3 stats logging with vanilla_stats in
the manycore

This commit creates two sets of Counters (Epoch and Tag) and associated Printing
Functions (PrintEpochStats and PrintTagStats) within each class
memory_system->dram_system->controller->simple_stats

Within each set there are three types:
{set}_counters_ {set}_vec_counters {set}_histo_counts_

The only difference between the two sets of Counters is the period when they
print to a file and reset their counters

Epoch - prints every "epoch_period" cycle which is set in each dramsim3 config
file after it prints a period, the stats are accumulated in the total counter
and then reset to 0

Tag - prints whenever "print_stat_v_i" in manycore_tb_top is set high, along
with the associated tag in "print_stat_tag_i". This is the same with how
vanilla_stats.csv gets logged. After it prints the counter is reset to 0, but
they are not accumulated.
```

drichmond commented:
Would be good to put that in the source somewhere, too.

marcusnchow commented:
added message in simple_stats.h

drichmond commented:
> added message in simple_stats.h

I don't see it?
